### ISSUE TITLE
Confirm you are a human

### ISSUE THREAD

**Issue by nattzp** [jonkpirateboy/Pass-fur-alle#42]
They just added a text field where you have to type in letters from a picture to confirm you are a human. Is there any way around this? Maybe edit the script so it starts running after you have typed in the letters, which you only have to do one time in the beginning?

**Comment by barhom** [comment]
this  confirmation was just added

**Comment by geekteq** [comment]
I saw it on the confirmation page as well but it may have been moved.

**Comment by geekteq** [comment]
Confirmed it's on the confirmation page as well.
So I'm assuming we need to manually confirm as well

https://capture.dropbox.com/rGqRSfwjx9ezcM18

**Comment by haegg78** [comment]
They have also blocked number of requests now
<img width="903" alt="Skärmavbild 2022-03-31 kl  17 09 58" src="https://user-images.githubusercontent.com/6528200/161088854-1927553b-6a8a-46d3-9dac-dddfeb22835e.png">

**Comment by nattzp** [comment]
They also added a confirmation captcha on the "confirm booking" page as well

**Comment by nattzp** [comment]
> 

ooops sry hehe

**Comment by geekteq** [comment]
> They have also blocked number of requests now <img alt="Skärmavbild 2022-03-31 kl 17 09 58" width="903" src="https://user-images.githubusercontent.com/6528200/161088854-1927553b-6a8a-46d3-9dac-dddfeb22835e.png">

I get through this blockage by using Firefox instead of Chrome so I'm assuming they have something that's set for us to block the user agent.

**Comment by justinphelps** [comment]
Can confirm that using the Firefox/geckdriver instead of Chrome fixes the "För många anprop" issue, but the Captcha issue still blocks .

**Comment by addellll** [comment]
Just to add some value for this problem maybe Deathbycaptcha/Anti-Captcha paid API can solve this? 

Big thank you @jonkpirateboy for this tool. And it seams that I was lucky since I got time 2 days ago before this captcha BS.

**Comment by d00ze** [comment]
@jonkpirateboy Kämpa på, jag han fick lägga till en funktion för paus i ditt tidigare script. Men fick en tid idag och passet är fixat.
Tack för ditt arbete.

**Comment by Leopenardo** [comment]
> @jonkpirateboy Kämpa på, jag han fick lägga till en funktion för paus i ditt tidigare script. Men fick en tid idag och passet är fixat. Tack för ditt arbete.

Skulle du snälla kunna hjälpa mig?

**Comment by kajetan-kazimierczak** [comment]
* Disable javascript to make the captcha go away.
* Stop DDOSing the poor booking site (set time between calls to 10-15 seconds or longer) to avoid the "för många anrop"

**Comment by andreasmq** [comment]
faking useragent with

```
ua = UserAgent()
user_agent = ua.ff
options.add_argument(f'user-agent={user_agent}')
```

does results in the "för många anrop" page 

Switching to firefox driver instead of chrome gets to the captcha.

**Comment by LucGosso** [comment]
Polisens captcha, detta var lite ironiskt

![image](https://user-images.githubusercontent.com/9716195/161150356-b462672d-72b0-4d8c-a76d-ee0dbc97d41c.png)

**Comment by nattzp** [comment]
> * Disable javascript to make the captcha go away.
> * Stop DDOSing the poor booking site (set time between calls to 10-15 seconds or longer) to avoid the "för många anrop"

how do i set the time between calls?

**Comment by nattzp** [comment]
Could we instead have an input for manually writing the captcha letters into the terminal? How would that be implemented with code?

**Comment by nattzp** [comment]
Solved the captcha issue by adding a sleep(10) before continuing to next page on the captcha page, allows you to manually insert captcha letters. However, it seems that they have changes the layout for the times available now so the code doesn't work anyway. Could this be updated to match the new layout?

**Comment by UgisL** [comment]
For me, it seems that the time choice works only with javascript turned on ... Would it be possible to dynamically turn on / off the javascript permission?

I.e., start with turned off (first page), when we are at checkpoint that we live in Sweden, turn it back on, then after the time choice turn it off (need to press next 2 times, because after the choice of time, the captcha is immediately loaded and first time around it complains about not filling it in ...

When the script was working, I have been searching for time further away from now (Skatteverket has not registered the name of my newborn just yet, so I can not rush to polisen any time now), between beginning of May and mid-June, was not lucky yet.

**Comment by Andows** [comment]
It is also a captcha on the confirm booking page now.

**Comment by mryner228** [comment]
ok, so I seem to be able to get the script running in firefox. Set a delay to be able to fill in captcha on the first page. First available time seems to be an issue, that just halts. Setting 'firstDate' to false starts the search at least.

With that set to False I manage to go through to the booking. Tried with dates in Oct (so way late but just to try). But for me everything except 'firstDate' seems to work in firefox.

**Comment by mryner228** [comment]
With firstDate = true scripts only seems to run twice and then halt.

**Comment by geekteq** [comment]
> ok, so I seem to be able to get the script running in firefox. Set a delay to be able to fill in captcha on the first page. First available time seems to be an issue, that just halts. Setting 'firstDate' to false starts the search at least.
> 
> With that set to False I manage to go through to the booking. Tried with dates in Oct (so way late but just to try). But for me everything except 'firstDate' seems to work in firefox.

Do share the code if you want to. Have my last 4 people to get through and even if it's manual labor I'd like to help them out ;)

**Comment by mryner228** [comment]
Haven't changed any code really. Just called on Firefox instead of Chrome with Gecko. Then added a delay on the first page so you have time to enter the captcha letters. After that just let it roll. Not as effective as firstDate but. 

This is only change in code really apart from calling on firefox. So time.sleep(10) instead of (1)

     # Click the first button
        startButton = web.find_element(by=By.XPATH, value='//*[@class="btn btn-primary btn-large disable-on-save"]')
        startButton.click()
        time.sleep(10)

**Comment by mryner228** [comment]
Seem to drop after a couple of loops though... not sure why and that seems to happen after 2 loops on 'firstDate' instead of 5-6 going through a range of dates...

**Comment by fltman** [comment]
So I changed the script to use firefox. 

`web = webdriver.Firefox("/Applications/Firefox.app/Contents/MacOS/")
`
Then, before entering the location I went to **about:config** in the firefox window, searched for javascript and disabled it. 
And it just worked... for about 20 minutes or so then firefox was blocked (to many calls).

I can still access the site from other browsers and devices so the lock isn't based on IP alone. Cleaning the cookies and changing the useragent didn't do it either. This means that the to many calls block must be based on more things. 

Perhaps trying to run the session through a TOR proxy in combination with the things above could bypass it?

**Comment by fltman** [comment]
Nope, didn't work through a TOR proxy either. So they detect and log something else about the client.

**Comment by MarcusWidle** [comment]
> Just to add some value for this problem maybe Deathbycaptcha/Anti-Captcha paid API can solve this?
> 
> Big thank you @jonkpirateboy for this tool. And it seams that I was lucky since I got time 2 days ago before this captcha BS.

Anti-Captcha or any other services wouldn't really help as none of them are supports mtcaptcha.

**Comment by ispjstfu** [comment]
I just made it work by adding the following code after the #Accept-section in the code

`        # MANUALLY Solve Captcha
        input("Lös verifieringen med Captcha manuellt och klicka på enter när du är klar.")
        manualverifyNextButton = web.find_element(by=By.XPATH, value='//*[@id="Main"]/form/div[2]/input')
        manualverifyNextButton.click()
        time.sleep(1)`

This stops the script indefinitely so we can input the captcha manually, and then we can just hit enter when ready to move on. Works perfectly for me.

**Comment by tedpelas** [comment]
Anyone else also been getting these errors recently?
```
504 Gateway Time-out
The server didn't respond in time.
```

**Comment by danwie** [comment]
> Can confirm that using the Firefox/geckdriver instead of Chrome fixes the "För många anprop" issue, but the Captcha issue still blocks .

Could you please share the code needed to use Firefox/geckdriver?

**Comment by rypto** [comment]
> 

Do you replace some other lines in the code then or how do you do it?, I'm running Ubuntu and have installed geckodriver

**Comment by rypto** [comment]
> So I changed the script to use firefox.
> 
> `web = webdriver.Firefox("/Applications/Firefox.app/Contents/MacOS/") ` Then, before entering the location I went to **about:config** in the firefox window, searched for javascript and disabled it. And it just worked... for about 20 minutes or so then firefox was blocked (to many calls).
> 
> I can still access the site from other browsers and devices so the lock isn't based on IP alone. Cleaning the cookies and changing the useragent didn't do it either. This means that the to many calls block must be based on more things.
> 
> Perhaps trying to run the session through a TOR proxy in combination with the things above could bypass it?

Do you replace some other lines in the code then or how do you do it?, I'm running Ubuntu and have installed geckodriver

**Comment by geekteq** [comment]
> Now I've also added a fix for the second verification step after a time is found, plus a click that the "important information" was read so that the script can move on and confirm the booking.

Thanks, Running with these fixes now but also added so I have to click the "boka ny tid" knappen manuellt för de verkade känna av om den klickades "programmatiskt" eller inte och skickade mig till "för många anrop" - klickar jag den själv så kommer jag förbi.

Fyller Captchan och sen åker jag. Eftersom det kan ta en stund så tankade jag in "playsound" när det är dags för sista captchan också. Skrämde frun när en "aaaauuuugahhhh" spelades två gånger igår kväll - hann fram, körde igenom captchan så nu har jag bara två kompisar kvar på listan innan jag kan strunta i det här :)

**Comment by danwie** [comment]
> > Now I've also added a fix for the second verification step after a time is found, plus a click that the "important information" was read so that the script can move on and confirm the booking.
> 
> Tack, kör med dessa nu men la till "boka ny tid" knappen manuellt för de verkade känna av om den klickades "programmatiskt" eller inte och skickade mig till "för många anrop" - klickar jag den själv så kommer jag förbi.
> 
> Fyller Captchan och sen åker jag. Eftersom det kan ta en stund så tankade jag in "playsound" när det är dags för sista captchan också. Skrämde frun när en "aaaauuuugahhhh" spelades två gånger igår kväll - hann fram, körde igenom captchan så nu har jag bara två kompisar kvar på listan innan jag kan strunta i det här :)

Tack - det var den informationen jag behövde. Nu snurrar det hos mig här också. Men spela upp ljud när sista captchan dyker upp, hur fixade du det?

**Comment by geekteq** [comment]
> Tack - det var den informationen jag behövde. Nu snurrar det hos mig här också. Men spela upp ljud när sista captchan dyker upp, hur fixade du det?

Litet hack.. 
I början hämtar jag in 

`from playsound import playsound` (behöver köra "pip3 install playsound" före)

Vid den sista "captchan"

`            # MANUALLY solve verification again
            input("Lös även denna verifieringen manuellt och klicka på enter när du är klar.")
            manualverify2NextButton = web.find_element(by=By.XPATH, value='//*[@id="Main"]/form/div[2]/input')
            manualverify2NextButton.click()
            time.sleep(1)`

så lägger jag in

`            # MANUALLY solve verification again
            playsound('warning.mp3') #vilken mp3a som helst funkar o ska ligga i samma bibliotek
            input("Lös även denna verifieringen manuellt och klicka på enter när du är klar.")
            manualverify2NextButton = web.find_element(by=By.XPATH, value='//*[@id="Main"]/form/div[2]/input')
            manualverify2NextButton.click()
            time.sleep(1)`

**Comment by dio99** [comment]
tjo testar detta hack men får endå upp den bilden när man ska clicka på boka tid och när ja klickar på den så blir det den blocken att man försöker för ofta

**Comment by geekteq** [comment]
> @geekteq tjo testar detta hack men får endå upp den bilden när man ska clicka på boka tid och när ja klickar på den så blir det den blocken att man försöker för ofta

"Works for me" - men kör Mac-lösning så vet inte vad som kan diffa. Uppdaterat allt chrome, chromedriver etc igår så kan ha med saken att göra antar jag - men vet inte..

**Comment by dio99** [comment]
> > @geekteq tjo testar detta hack men får endå upp den bilden när man ska clicka på boka tid och när ja klickar på den så blir det den blocken att man försöker för ofta
> 
> "Works for me" - men kör Mac-lösning så vet inte vad som kan diffa. Uppdaterat allt chrome, chromedriver etc igår så kan ha med saken att göra antar jag - men vet inte..

hmm linux testar ja med.. ok ska se om ja kan dra ner nyare driver då för chrome. kör default deb paket

**Comment by UgisL** [comment]
Finally, got also our time. I used Chrome, with these load options:
option = webdriver.ChromeOptions()
option.add_argument('--disable-blink-features=AutomationControlled')
web = webdriver.Chrome(executable_path='/usr/bin/chromedriver',options=option)

And commented out in the script automatic click of the first button, and also had to manually enter both captchas. I used 20s sleep between different "Find first available time" clicks. Then I implemented email notification, when I had to run to the computer and enter the final captcha and finish the booking. Email call looks like (will stop working with google after end of May, will have to use more complex authentification):
import smtplib, ssl
port = 465  # For SSL
password = "test@1234"
# Create a secure SSL context
context = ssl.create_default_context()

with smtplib.SMTP_SSL("smtp.gmail.com", port, context=context) as server:
                server.login("[yourmail]@gmail.com", password)
                server.sendmail("[yourmail2]@gmail.com", "[yourmail2]@gmail.com", """\
Subject: Finish passport booking!

I found a good time!!!""")

Good luck, everyone!

**Comment by geekteq** [comment]
@UgisL nice one. Makes me thing I should've gone with "pushover" rather than "playsound" but only have one friend more to help so can't do it. 

@jonkpirateboy i don't know if it's beer-ware but do let me know how I can thank you with more than "thank you" for these hacks now that I'm almost done 🥇

**Comment by dio99** [comment]
> Finally, got also our time. I used Chrome, with these load options: option = webdriver.ChromeOptions() option.add_argument('--disable-blink-features=AutomationControlled') web = webdriver.Chrome(executable_path='/usr/bin/chromedriver',options=option)
> 
> And commented out in the script automatic click of the first button, and also had to manually enter both captchas. I used 20s sleep between different "Find first available time" clicks. Then I implemented email notification, when I had to run to the computer and enter the final captcha and finish the booking. Email call looks like (will stop working with google after end of May, will have to use more complex authentification): import smtplib, ssl port = 465 # For SSL password = "test@1234" context = ssl.create_default_context()
> 
> with smtplib.SMTP_SSL("smtp.gmail.com", port, context=context) as server: server.login("[yourmail]@gmail.com", password) server.sendmail("[yourmail2]@gmail.com", "[yourmail2]@gmail.com", """ Subject: Finish passport booking!
> 
> I found a good time!!!""")
> 
> Good luck, everyone!

cahnged like u dig and it seems running now, but in what page did u stop posting some info manually ?
im in sektion now where u choose sök tid

**Comment by BloodDragon05** [comment]
How do you guys get pass "Boka tid"-button? As soon as i press that button i get sent to "För många anrop"...

**Comment by geekteq** [comment]
> How do you guys get pass "Boka tid"-button? As soon as i press that button i get sent to "För många anrop"...

I don't know tbh. I click the button manually and it doesn't tell me "för många anrop", if I try it without doing it manually I get slapped in the face with "För många androp". As mentioned I have updated all my systems and I'm running MacOS but I don't know if that is the difference.

**Comment by BloodDragon05** [comment]
> > How do you guys get pass "Boka tid"-button? As soon as i press that button i get sent to "För många anrop"...
> 
> I don't know tbh. I click the button manually and it doesn't tell me "för många anrop", if I try it without doing it manually I get slapped in the face with "För många androp". As mentioned I have updated all my systems and I'm running MacOS but I don't know if that is the difference.

What version om ChromeDriver are you using? I am on windows and followed every step but maybe I've done something wrong... I press the button manually since the script won't press it...

**Comment by geekteq** [comment]
> What version om ChromeDriver are you using? I am on windows and followed every step but maybe I've done something 
> wrong... I press the button manually since the script won't press it...

100.0.4896.60

**Comment by nattzp** [comment]
Can someone who has working code publish their entire .py file? I am in desperate need hehe

**Comment by BloodDragon05** [comment]
> > What version om ChromeDriver are you using? I am on windows and followed every step but maybe I've done something
> > wrong... I press the button manually since the script won't press it...
> 
> 100.0.4896.60

Have that too... Get kicked out everytime....

**Comment by geekteq** [comment]
[Here's mine without "playsound"](https://www.dropbox.com/s/i8tlvw32usu04dj/pass-fur-alle.py?dl=0)

**Comment by jonkpirateboy** [comment]
Works great @geekteq! Do you want to add this as a Pull request?

**Comment by geekteq** [comment]
> Works great @geekteq! Do you want to add this as a Pull request?

My first pull request so don't be too harsh on me :)

**Comment by moyomo77** [comment]
Change
        ......... 
       infoCheck.click()
        infoNext.click()
        time.sleep(1)
        # MANUALLY solve verification
        input("Lös verifieringen manuellt och klicka på enter när du är klar.")
        manualverifyNextButton = web.find_element(by=By.XPATH, value='//*[@id="Main"]/form/div[2]/input')
        manualverifyNextButton.click()
       ......... 
with this, works prefecty for me, redhat desktop :)
         ......... 
        time.sleep(1)
        # MANUALLY solve verification
        input("Lös verifieringen manuellt och klicka på enter när du är klar.")
        manualverifyNextButton = web.find_element(by=By.XPATH, value='//*[@id="Main"]/form/div[2]/input')
        manualverifyNextButton.click()
        infoCheck.click()
        infoNext.click()
       .........

**Comment by geekteq** [comment]
A little wave to the nemoq devs, or consultants, that seem to be following the discussion and adapting along the way. This is my assumption due to the quick patching of our "circumvents". I get it and I don't blame you - just too bad that you didn't fix it before there are no times for 6 months (then this issue wouldn't have been a thing). My suggestion is "BankID" to book.. would be easist for all :)

**Comment by rypto** [comment]
> [Here's mine without "playsound"](https://www.dropbox.com/s/i8tlvw32usu04dj/pass-fur-alle.py?dl=0)

Downloaded this and ran it on my ubuntu desktop. I can confirm I get the "for manga anrop" after I get by the "Boka ny tid" button. It opens up and run with Chrome version 100.0.4896.60 (Official Build) (64-bit)

**Comment by ellen9898** [comment]
I couldn't get this project to work but I found a working one https://github.com/kalkih/passport-appointment-bot

Successfully booked a time with it!

**Comment by UgisL** [comment]
> changed like u did and it seems running now, but in what page did u stop posting some info manually ?
> im in sektion now where u choose sök tid

I had to stop the script where it fills in the name, last name, etc., because there is the second captcha there ... For that I followed one of the suggestions in this same discussion (in which python script waits for pressing a key to continue). Now, with the script I was using 2 days ago, I also get into "too many tries" webpage.

But this stripped down version still seems to work (notice that I have removed all personal info, and also the gmail account I was using - for gmail account, insecure communication was needed): (change extension from gz to py)
[pass-fur-alle_github.gz](https://github.com/jonkpirateboy/Pass-fur-alle/files/8407678/pass-fur-alle_github.gz)

Everythin is manual, except checking for free time every 20 seconds.
